### PR TITLE
:bug: Source Monday: fix pagination for the items stream

### PR DESCRIPTION
**Pull request by igsaf2**
## What
Relates: 
- https://github.com/airbytehq/airbyte/issues/25862
- https://github.com/airbytehq/airbyte/issues/25683

Two issues:
- no pagination for `items` inside each `board` (with the current limit of 100 per `board`)
- stop iteration through pages when a `board` without `items` encountered
 
In more details see this comment: https://github.com/airbytehq/airbyte/issues/25683#issuecomment-1536181166

## How
Created a new pagination strategy for `items` stream that allow iteration through both, `boards` and `items`.
Suppose we have 4 boards (their items are 10, 150, 0, 20 respectively)
In the **current** version:
```
boards(limit:1){items(limit:100){assets{id}...
boards(limit:1,page:2){items(limit:100){assets{id}...  # 100 out of 150 items fetched
boards(limit:1,page:3){items(limit:100){assets{id}...
# the third board has no items, the iteration is completed
```
In the **PR** (`items(limit:100,page:1)` behaves in the same way as `items(limit:100)`:
```
boards(limit:1){items(limit:100){assets{id}...
boards(limit:1,page:2){items(limit:100,page:1){assets{id}...  # fetched 100 items for board 2
boards(limit:1,page:2){items(limit:100,page:2){assets{id}...  # fetched remaining 50 items for board 2
boards(limit:1,page:3){items(limit:100,page:1){assets{id}...  # no items, but keep going to the next board
boards(limit:1,page:4){items(limit:100,page:1){assets{id}...
```

## Recommended reading order
1. `airbyte-integrations/connectors/source-monday/unit_tests/test_graphql_requester.py`
2. `airbyte-integrations/connectors/source-monday/unit_tests/test_item_pagination_strategy.py`
3. `airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py`
4. `airbyte-integrations/connectors/source-monday/source_monday/graphql_requester.py`
5. `airbyte-integrations/connectors/source-monday/source_monday/manifest.yaml`

## 🚨 User Impact 🚨
No breaking changes

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

### CODE REVIEW

**Comment by igsaf2** [comment]
Tested it with `python main.py read --config secrets/config.json --catalog configured_stream.json` and was able to get 4000+ items instead of 400 before.

File: airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                return None
+
+        return self._page, self._sub_page
+
+    def _can_board_have_more_items(self, last_records):
+        return len(last_records) >= self.page_size
+
+    @staticmethod
+    def _can_have_more_boards(response):
+        return bool(len(response.json()["data"]["boards"]))
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py:41]: Are you using this function in other place? If not move the code the function itself.

File: airbyte-integrations/connectors/source-monday/source_monday/manifest.yaml [hunk 2]
```diff
@@ -63,8 +63,10 @@ definitions:
       paginator:
         $ref: "#/definitions/default_paginator"
         pagination_strategy:
-          $ref: "#/definitions/default_paginator/pagination_strategy"
-          page_size: 1
+          class_name: "source_monday.item_pagination_strategy.ItemPaginationStrategy"
+          type: "CustomPaginationStrategy"
+          page_size: 100
+          start_from_page: 1
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-monday/source_monday/manifest.yaml:69]: This is not clear the page and subpage will start from 1. Any reason why you'd like to have control over this parameter?

File: airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+from typing import Any, Mapping, Optional, List, Tuple
+
+from airbyte_cdk.sources.declarative.requesters.paginators.strategies.page_increment import PageIncrement
+
+
+class ItemPaginationStrategy(PageIncrement):
+    """
+    Page increment strategy with subpages for the `items` stream
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py:8]: @igsaf2 can you add link to Monday API documentation page here?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-monday/source_monday/manifest.yaml (hunk 2 above)
Comment (igsaf2) [reply airbyte-integrations/connectors/source-monday/source_monday/manifest.yaml:69]: Both :-)
I will introduce another parameter for sub-page

Comment (igsaf2) [reply airbyte-integrations/connectors/source-monday/source_monday/manifest.yaml:69]: You are right. With a custom pagination strategy, this parameter makes little sense. I will embed this value into the strategy.

**Comment by igsaf2** [comment]
@marcosmarxm I fixed the PR, could you please review again?

File: airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+from typing import Any, Mapping, Optional, List, Tuple
+
+from airbyte_cdk.sources.declarative.requesters.paginators.strategies.page_increment import PageIncrement
+
+
+class ItemPaginationStrategy(PageIncrement):
+    """
+    Page increment strategy with subpages for the `items` stream.
+
+    See Monday API documentation https://developer.monday.com/api-reference/docs.
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py:10]: Can you reference a better location where future developers can find information to this specific custom component?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py (hunk 4 above)
Comment (igsaf2) [reply airbyte-integrations/connectors/source-monday/source_monday/item_pagination_strategy.py:10]: I can put something like this:

From the `items` documentation https://developer.monday.com/api-reference/docs/items
> Please note that you cannot return more than 100 items per query when using items at the root. To [adjust](https://developer.monday.com/api-reference/changelog/new-limits-to-the-items-query-in-90-days) your query, try only returning items on a specific board, nesting items inside a boards query, looping through the boards on your account, or querying less than 100 items at a time.

This pagination strategy supports nested loop through `boards` on the top level and `items` on the second. See `boards` documentation for more details: https://developer.monday.com/api-reference/docs/boards#queries

**Comment by igsaf2** [comment]
@marcosmarxm fixed

**Review by archangelic — APPROVED** [review]

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**